File: csharp_new/src/Vista.SDK/Gmod.cs
using System.Collections;
using System.Diagnostics.CodeAnalysis;
using uniffi.bindings;
using Vista.SDK.Internal;

namespace Vista.SDK;

public delegate TraversalHandlerResult TraverseHandler(IReadOnlyList<GmodNode> parents, GmodNode node);

public sealed class Gmod
{
    private class TraverseCallbackImpl : TraversalCallback
    {
        private readonly TraverseHandler _handler;

        public TraverseCallbackImpl(TraverseHandler handler) => _handler = handler;

        public TraversalHandlerResult Handler(List<uniffi.bindings.GmodNode> parents, uniffi.bindings.GmodNode node) =>
            _handler(parents.Select(GmodNode.FromBindings).ToList(), node.ToGmodNode());
    }

    private readonly uniffi.bindings.Gmod _inner;
    public VisVersion VisVersion => _inner.Version();

    private Gmod(uniffi.bindings.Gmod inner) => _inner = inner;

    internal static Gmod FromBindings(uniffi.bindings.Gmod inner) => new Gmod(inner);

    public bool Traverse(TraverseHandler handler) => _inner.Traverse(new TraverseCallbackImpl(handler));

    public GmodNode this[string key] => _inner.GetNode(key).ToGmodNode();

    public bool TryGetNode(string code, [MaybeNullWhen(false)] out GmodNode node)
    {
        node = default;
        var result = _inner.TryGetNode(code);
        if (result is null)
            return false;
        node = result.ToGmodNode();
        return true;
    }
}

File: rust/bindgen/src/gmod.rs
use std::{fmt::Debug, sync::Arc};

use uniffi::deps::log::error;
use vista_sdk::{gmod::TraversalHandlerResult, vis::VisVersion};

use crate::{internal::common::Exposable, Gmod, GmodNode};

impl Gmod {
    pub fn version(&self) -> VisVersion {
        self.inner.version()
    }

    pub fn root_node(&self) -> Arc<GmodNode> {
        self.inner.root_node().to_threadsafe_exposed()
    }

    pub fn get_node(&self, code: &str) -> Arc<GmodNode> {
        self.inner.get_node(code).to_threadsafe_exposed()
    }

    pub fn try_get_node(&self, code: &str) -> Option<Arc<GmodNode>> {
        match self.inner.try_get_node(code) {
            Some(inner_node) => Some(inner_node.to_threadsafe_exposed()),
            None => None,
        }
    }

    pub fn get_parents(&self, node: &GmodNode) -> Vec<Arc<GmodNode>> {
        self.inner
            .get_parents(&node.inner)
            .map(|node| node.to_threadsafe_exposed())
            .collect()
    }

    pub fn traverse(&self, callback: Box<dyn TraversalCallback>) -> bool {
        self.inner.traverse(|_parents, _node| {
            callback
                .handler(
                    _parents.map(|node| node.to_threadsafe_exposed()).collect(),
                    _node.to_threadsafe_exposed(),
                )
                .unwrap()
        })
    }
}

pub(crate) trait TraversalCallback: Send + Sync + Debug {
    fn handler(
        &self,
        parents: Vec<Arc<GmodNode>>,
        node: Arc<GmodNode>,
    ) -> Result<TraversalHandlerResult, TraversalError>;
}

#[derive(Debug, thiserror::Error)]
pub(crate) enum TraversalError {
    #[error("Failed")]
    Failed,
}
